File: minimum-depth-of-binary-tree.cpp
//Runtime: 6 ms
class Solution {
public:
    int ans = INT_MAX;

    void traverse(TreeNode* root,int d)
    {
        if(!root)
            return;

        if(!root->left && !root->right)
            ans = min(ans, d);

        traverse(root->left, d+1);
        traverse(root->right, d+1);
    }

    int minDepth(TreeNode* root) {
        if(!root)
            return 0;

        traverse(root, 1);

        return ans;
    }
};

File: minimum-depth-of-binary-tree.scala
//Runtime: 707 ms
object Solution {
    def minDepth(root: TreeNode): Int = {
        if(root == null) 0
        else if(root.left == null && root.right == null) 1
        else if(root.left == null) 1 + minDepth(root.right)
        else if(root.right == null) 1 + minDepth(root.left)
        else 1 + Math.min(minDepth(root.right), minDepth(root.left))
    }
}
